### PR TITLE
NLP transform improvements

### PR DESCRIPTION
**Pull request by christhetree**
to: @asaluja @jq @deerzq 
cc: @hectorgon @yolken 

This PR contains the following improvements and implemented feature requests for several string / NLP related transforms:
- Create abstract StringTransform class for transforms that simply process all strings in a string feature and output a new string transform or overwrite the input string transform (NormalizeUtf8Transform, ConvertStringCaseTransform, ReplaceAllStringsTransform). This reduces unnecessary duplicate code and logic.
- Rename DeleteStringFeatureColumnTransform to DeleteStringFeatureFamilyTransform for consistency and clarity.
- Make specifying keys for the MoveFloatToStringTransform optional in which case all keys are moved. This is necessary for transforms that output float features with an arbitrary amount of different keys (DefaultStringTokenizerTransform does this).
- Make the bigram flag optional in DefaultStringTokenizer in preparation for the ngram transform that will be added soon, at which point the bigram ability of the default string tokenizer transform will become obsolete.
- Add the ability to delete multiple string and float families with the DeleteStringFeatureFamilyTransform and DeleteFloatFeatureFamilyTransform in order to help cut down on the length of NLP related config files.
- Preserving backwards compatibility for all of these changes.
- Updating all related tests for these changes.

### CODE REVIEW

File: core/src/main/java/com/airbnb/aerosolve/core/transforms/StringTransform.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
+package com.airbnb.aerosolve.core.transforms;
```
Comment (jq) [inline core/src/main/java/com/airbnb/aerosolve/core/transforms/StringTransform.java]: move this to types?

File: core/src/main/java/com/airbnb/aerosolve/core/transforms/ConvertStringCaseTransform.java [hunk 2]
```diff
@@ -1,52 +1,38 @@
...
 /**
  * Converts strings to either all lowercase or all uppercase
  * "field1" specifies the key of the feature
  * "convert_to_uppercase" converts strings to uppercase if true, otherwise converts to lowercase
+ * "output" optionally specifies the key of the output feature, if it is not given the transform
+ * overwrites / replaces the input feature
  */
-public class ConvertStringCaseTransform implements Transform {
-  private String fieldName1;
+public class ConvertStringCaseTransform extends StringTransform {
```
Comment (jq) [inline core/src/main/java/com/airbnb/aerosolve/core/transforms/ConvertStringCaseTransform.java]: nice we need more abstract transform like StringTransform

File: core/src/main/java/com/airbnb/aerosolve/core/transforms/TransformFactory.java [hunk 3]
```diff
@@ -15,6 +15,12 @@ public static Transform createTransform(Config config, String key) {
     if (transformName == null) {
       return null;
     }
+
+    // TODO: remove this once all configs have migrated over to the new transform names
+    if (transformName.equals("delete_string_feature_column")) {
```
Comment (jq) [inline core/src/main/java/com/airbnb/aerosolve/core/transforms/TransformFactory.java]: you can keep the class just extend new class, this achieve the same goal.

File: core/src/main/java/com/airbnb/aerosolve/core/transforms/MoveFloatToStringTransform.java [hunk 4]
```diff
@@ -48,17 +52,37 @@ public void doTransform(FeatureVector featureVector) {
...
+      String key,
+      double cap,
+      double bucket) {
+    if (feature1.containsKey(key)) {
+      Double dbl = feature1.get(key);
+      if (dbl > cap) {
+        dbl = cap;
+      }
+
+      Double quantized = LinearLogQuantizeTransform.quantize(dbl, bucket);
```
Comment (deerzq) [inline core/src/main/java/com/airbnb/aerosolve/core/transforms/MoveFloatToStringTransform.java]: we discussed about how confusing this is to have quantize in `LinearLogQuantizeTransform`, perhaps you can refactor this part in a future PR

File: core/src/main/java/com/airbnb/aerosolve/core/transforms/DeleteFloatFeatureFamilyTransform.java [hunk 5]
```diff
@@ -1,19 +1,29 @@
...
 
   @Override
   public void configure(Config config, String key) {
-    fieldName1 = config.getString(key + ".field1");
+    if (config.hasPath(key + ".field1")) {
+      fieldName1 = config.getString(key + ".field1");
+    }
+    if (config.hasPath(key + ".fields")) {
+      fieldNames = config.getStringList(key + ".fields");
+    }
```
Comment (deerzq) [inline core/src/main/java/com/airbnb/aerosolve/core/transforms/DeleteFloatFeatureFamilyTransform.java]: If there is no existing production job relying on `field1`, it seems we can get rid of it. When the user want to delete just one family, just put it in `fields` would be ok.

Comment (christhetree) [inline core/src/main/java/com/airbnb/aerosolve/core/transforms/DeleteFloatFeatureFamilyTransform.java]: Done.

File: core/src/main/java/com/airbnb/aerosolve/core/transforms/MoveFloatToStringTransform.java (hunk 4 above)
Comment (christhetree) [inline core/src/main/java/com/airbnb/aerosolve/core/transforms/MoveFloatToStringTransform.java]: Sorry, I do have a PR almost ready to go for this, the NLP stuff just was more urgent. I'll push it up ASAP!

File: core/src/main/java/com/airbnb/aerosolve/core/transforms/TransformFactory.java (hunk 3 above)
Comment (christhetree) [inline core/src/main/java/com/airbnb/aerosolve/core/transforms/TransformFactory.java]: Done.

File: core/src/main/java/com/airbnb/aerosolve/core/transforms/ConvertStringCaseTransform.java (hunk 2 above)
Comment (christhetree) [inline core/src/main/java/com/airbnb/aerosolve/core/transforms/ConvertStringCaseTransform.java]: Thanks! :)

File: core/src/main/java/com/airbnb/aerosolve/core/transforms/StringTransform.java (hunk 1 above)
Comment (christhetree) [inline core/src/main/java/com/airbnb/aerosolve/core/transforms/StringTransform.java]: Done.

**Comment by christhetree** [comment]
@jq @deerzq thanks for the comments, all have been addressed, PTAL

**Comment by jq** [comment]
👍

**Pull request merged**